## Week 5 task by Denis Khryashchev

In [75]:
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans
from matplotlib import pyplot as plt
from scipy.cluster.hierarchy import dendrogram, linkage

pd.set_option("display.max_colwidth", 10)
pd.options.display.max_columns = 800
%matplotlib inline

### 0. Accessing the MNIST dataset

In [2]:
CSV_URL = "https://pjreddie.com/media/files/mnist_train.csv"

In [3]:
# Loading the entire dataset
data = pd.read_csv(CSV_URL, header = None)

In [6]:
data.head(3)

0    1    2    3    4    5    6    7    8    9    10   11   12   13   14   \
0    5    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    4    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    1    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    9    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   15   16   17   18   19   20   21   22   23   24   25   26   27   28   29   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   30   31   32   33   34   35   36   37   38   39   40   41   42   43   44   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   45   46   47   48   49   50   51   52   53   54   55   56   57   58   59   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   60   61   62   63   64   65   66   67   68   69   70   71   72   73   74   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   75   76   77   78   79   80   81   82   83   84   85   86   87   88   89   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   90   91   92   93   94   95   96   97   98   99   100  101  102  103  104  \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   105  106  107  108  109  110  111  112  113  114  115  116  117  118  119  \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
3    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
4    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   120  121  122  123  124  125  126  127  128  129  130  131  132  133  134  \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    

In [10]:
# Dropping the labels for now
labels = list(data[data.columns[0]])
data = data.drop([data.columns[0]], axis=1)

In [17]:
data.head(3)

1    2    3    4    5    6    7    8    9    10   11   12   13   14   15   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   16   17   18   19   20   21   22   23   24   25   26   27   28   29   30   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   31   32   33   34   35   36   37   38   39   40   41   42   43   44   45   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   46   47   48   49   50   51   52   53   54   55   56   57   58   59   60   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   61   62   63   64   65   66   67   68   69   70   71   72   73   74   75   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   76   77   78   79   80   81   82   83   84   85   86   87   88   89   90   \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   91   92   93   94   95   96   97   98   99   100  101  102  103  104  105  \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   106  107  108  109  110  111  112  113  114  115  116  117  118  119  120  \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   121  122  123  124  125  126  127  128  129  130  131  132  133  134  135  \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0   51  159  253  159   50    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   136  137  138  139  140  141  142  143  144  145  146  147  148  149  150  \
0    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   

   151  152  153  154  155  156  157  158  159  160  161  162  163  164  165  \
0    0    0    3   18   18   18  126  136  175   26  166  255  247  127    0   
1    0    0    0    0   48  238  252  252  252  237    0    0    0    0    0   
2    0    0    0    0    0    0    0    0    0    0   67  232   39    0    0   

   166  167  168  169  170  171  172  173  174  175  176  177  178  179  180  \
0    0    0    0    0    0    0    0    0    0    0    0   30   36   94  154   
1    0    0    0    0    0    0    0    0    0    0    0    0    0    0    0   
2    0    0    0    0    0    0    0   62   81    0    0    0    0    0    0   

   181  182  183  184  185  186  187  188  189  190  191  192  193  194  195  \
0  170  253  253  253  253  253  225  172  253  242  195   64    0    0

In [22]:
# Creating the vectors
X = np.asarray(data.values.tolist())

### 1. K-means clustering, 10 clusters

In [32]:
clusters = KMeans(n_clusters = 10, random_state = 2926).fit(X)

In [37]:
# Extracting the cluster labels
cluster_labels = list(clusters.labels_)

In [41]:
# Identifying unique true labels

In [52]:
def get_label_count(cluster_number, true_labels, cluster_labels):
    return sorted(zip(range(10), 
            np.bincount([labels[i] for i in range(len(cluster_labels)) if cluster_labels[i] == cluster_number])), 
                  key = lambda x: x[1], reverse = True)

In [56]:
# Printing the counts for the labels in each of the clusters
for cluster_number in range(10):
    print "Cluster # " + str(cluster_number), get_label_count(cluster_number, labels, cluster_labels)

Cluster # 0 [(1, 2977), (5, 801), (8, 421), (2, 352), (4, 309), (7, 283), (6, 224), (9, 117), (3, 107), (0, 21)]
Cluster # 1 [(4, 3188), (9, 2902), (7, 1795), (5, 376), (8, 196), (3, 175), (2, 173), (6, 82), (0, 38), (1, 6)]
Cluster # 2 [(0, 2499), (5, 274), (3, 127), (6, 122), (2, 98), (8, 32), (9, 19), (7, 14), (4, 12), (1, 0)]
Cluster # 3 [(1, 3718), (3, 426), (2, 363), (7, 327), (8, 297), (6, 271), (9, 240), (5, 167), (4, 159), (0, 3)]
Cluster # 4 [(6, 4914), (2, 212), (0, 182), (4, 165), (5, 122), (3, 55), (8, 47), (1, 8), (9, 8), (7, 4)]
Cluster # 5 [(0, 2810), (6, 99), (5, 62), (8, 38), (9, 37), (3, 15), (7, 15), (2, 11), (4, 10), (1, 0)]
Cluster # 6 [(7, 3773), (9, 2460), (4, 1942), (5, 353), (8, 179), (2, 69), (3, 48), (0, 14), (1, 9), (6, 1)]
Cluster # 7 [(2, 4194), (3, 219), (6, 86), (8, 54), (7, 39), (4, 37), (0, 17), (5, 14), (9, 13), (1, 9)]
Cluster # 8 [(8, 3463), (5, 1482), (3, 1036), (0, 177), (2, 158), (6, 90), (9, 69), (4, 19), (1, 10), (7, 10)]
Cluster # 9 [(3, 3923

## Summarizing the results, I can say that there are several clusters that accurately group the same numbers together. They are: 
### cluster # 7 that contains 4194 digits '2' out of 4633 (90+%) 
### cluster # 4 that contains 4914 digits '6' out of 5717 (85.9%)
## Other clusters have a relative majority of a specific digit that doesn't appear in another clusters:
### cluster # 8 that contains 3463 digits '8' out of 6514 (53.1%) followed by 1482 digits '5'
### cluster # 9 that contains 3923 digits '3' out of 7431 (52.8%) followed by 1770 digits '5'
### cluster # 6 that contains 3773 digits '7' out of 8848 (42.6%) followed by 2460 digits '9'
### cluster # 1 that contains 3188 digits '4' out of 8931 (35.7%) followed by 2902 digits '9'
## The left-over clusters # 0 and # 3 share the same digit '1' and clusters # 2 and # 5 share digit '0'. Even though they completely dominate in each of the clusters they are not unique overall. Therefore, using the 10-means clustering with default parameters I was not able to label the clusters with unique digits.

In [73]:
# Total counts for the labels
sorted(zip(range(10), np.bincount(labels)), key = lambda x: x[-1])

[(5, 5421),
 (4, 5842),
 (8, 5851),
 (6, 5918),
 (0, 5923),
 (9, 5949),
 (2, 5958),
 (3, 6131),
 (7, 6265),
 (1, 6742)]

### 2. Hierarchical clustering

In [ ]:
# Generating the linkage matrix with Ward
ward_linkage = linkage(X, 'ward')